### PR TITLE
ticket 6

### PR DESCRIPTION
**Pull request by iy2k22**
Made an attempt at Ticket 6.

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -104,18 +115,45 @@ describe("GET /api/reviews", () => {
...
+          owner: expect.any(String),
+          review_img_url: expect.any(String),
+          review_body: expect.any(String),
+          category: expect.any(String),
+          created_at: expect.any(String),
+          votes: expect.any(Number),
+        });
+      });
+  });
+  test("responds with 404 upon being a number out of range", () => {
```
Comment (robert112233) [inline __tests__/app.test.js]: This test does not test for a 404, it expects a 200 back

File: __tests__/app.test.js [hunk 2]
```diff
@@ -139,18 +177,63 @@ describe("GET /api/reviews/:review_id", () => {
...
\ No newline at end of file
+      .get("/api/reviews/a")
+      .expect(400)
+      .then(({ body }) => {
+        expect(body.msg).toBe("error: invalid input");
+      });
+  });
+});
+
+describe.only("GET /api/reviews/:review_id/comments", () => {
```
Comment (robert112233) [inline __tests__/app.test.js:184]: Remove .only !

File: __tests__/app.test.js [hunk 3]
```diff
@@ -104,18 +115,45 @@ describe("GET /api/reviews", () => {
...
+    return request(app).get("/api/reviews/1").expect(200);
+  });
+  test("responds with an object", () => {
+    return request(app)
+      .get("/api/reviews/1")
+      .then(({ body }) => {
+        expect(typeof body.review).toBe("object");
+      });
+  });
+  test("responds with object with appropriate properties", () => {
```
Comment (robert112233) [inline __tests__/app.test.js:132]: Isn't this test the same as 161 to 176 ? Why did you add it in again

File: __tests__/app.test.js [hunk 4]
```diff
@@ -139,18 +177,63 @@ describe("GET /api/reviews/:review_id", () => {
...
+
+describe.only("GET /api/reviews/:review_id/comments", () => {
+  test("returns status code 200", () => {
+    return request(app).get("/api/reviews/2/comments").expect(200);
+  });
+  test("responds with an object with a key of comments containing an array", () => {
+    return request(app)
+      .get("/api/reviews/2/comments")
+      .then(({ body }) => {
+        expect(body.comments).toBeInstanceOf(Array);
```
Comment (robert112233) [inline __tests__/app.test.js:192]: Its not really necessary to test that it's an array if you're using a forEach on the array later on, forEach only works on arrays so if line 211 passes you know you have an array. It's good that you considered it though!

File: __tests__/app.test.js [hunk 5]
```diff
@@ -139,18 +177,63 @@ describe("GET /api/reviews/:review_id", () => {
...
+            review_id: expect.any(Number),
+            created_at: expect.any(String),
+          });
+        });
+      });
+  });
+  test("if given a valid review id that doesn't have any comments, respond with 404 and an appropriate message", () => {
+    return request(app)
+      .get("/api/reviews/1/comments")
+      .expect(404)
```
Comment (robert112233) [inline __tests__/app.test.js]: This should be a 200. Nothing has gone wrong, it should return an empty array showing the user that there is no comments for that valid review 👍

File: __tests__/app.test.js [hunk 6]
```diff
@@ -139,18 +177,63 @@ describe("GET /api/reviews/:review_id", () => {
...
+      });
+  });
+  test("if given invalid id, respond with 400", () => {
+    return request(app)
+      .get("/api/reviews/a/comments")
+      .expect(400)
+      .then(({ body }) => {
+        expect(body.msg).toBe("error: invalid input");
+      });
+  });
```
Comment (robert112233) [inline __tests__/app.test.js:232]: You are missing a 404 test for invalid ID. Eg /api/reviews/384/comments 👍

File: controllers/index.js [hunk 7]
```diff
@@ -26,17 +27,22 @@ const getReviews = (req, res, next) => {
...
 const getReviewById = (req, res, next) => {
-    const id = req.params.review_id;
-    readReviewById(id).then((review) => {
+    readReviewById(req.params.review_id).then((review) => {
         res.status(200).send({ review: review });
     }).catch((err) => {
-        console.log(err);
         next(err);
     })
 }
```
Comment (robert112233) [inline controllers/index.js:36]: Here you will need to make a checkIfReviewExists function that allows you to differentiate between a valid review with no comments (200) and a non existent review (404). Both of these will return an empty array at this point. This function will make a query to the database, meaning you will need two queries in total. 

See these notes for assistance: 

https://notes.northcoders.com/courses/js-back-end/advanced-error-handling

File: __tests__/app.test.js (hunk 1 above)
Comment (robert112233) [reply __tests__/app.test.js]: I think you can probably delete it. It looks you have done sad path testing later on for this endpoint

**Review by robert112233 — CHANGES_REQUESTED** [review]
Good start but missing some tests. You also have some unnecessary code  / .only - remember to check through carefully before submitting a PR. 

Make adjustments then submit another PR for review 👍

File: __tests__/app.test.js [hunk 8]
```diff
@@ -56,9 +56,20 @@ describe("GET /api/reviews", () => {
...
+          "review_img_url",
+          "designer",
+          "review_id",
+          "votes",
+          "comment_count",
+          "created_at",
+        ];
         for (let review of res.body.reviews) {
-            for (let prop of props) expect(review.hasOwnProperty(prop)).toBe(true);
+          for (let prop of props)
```
Comment (roisinoreilly) [inline __tests__/app.test.js:71]: I'm not a fan of these names they're not the clearest

File: __tests__/app.test.js [hunk 9]
```diff
@@ -104,22 +115,37 @@ describe("GET /api/reviews", () => {
...
+  });
+  test("responds with object with appropriate properties", () => {
     return request(app)
-    .get("/api/reviews/1")
-    .expect(200);
-  })
+      .get("/api/reviews/1")
+      .then(({ body }) => {
+        expect(body.review).toMatchObject({
+          review_id: expect.any(Number),
```
Comment (roisinoreilly) [inline __tests__/app.test.js:137]: if we're only getting back the object of the review with id 1 we're not expecting any number we're expecting the review id to be 1

File: __tests__/app.test.js [hunk 10]
```diff
@@ -139,18 +165,69 @@ describe("GET /api/reviews/:review_id", () => {
...
-    .expect(400)
+      .get("/api/reviews/a")
+      .expect(400)
+      .then(({ body }) => {
+        expect(body.msg).toBe("error: invalid input");
+      });
+  });
+});
+
+describe.only("GET /api/reviews/:review_id/comments", () => {
```
Comment (roisinoreilly) [inline __tests__/app.test.js:184]: this still needs to be removed please

File: models/index.js [hunk 11]
```diff
@@ -17,8 +17,14 @@ const readReviewById = (id) => {
   .then((result) => result.rows[0] || Promise.reject({ status: 404, msg: "error: review id not found" }));
 }
 
+const readCommentsFromReview = (id) => {
+  return readReviewById(id).then(() => {
```
Comment (roisinoreilly) [inline models/index.js:21]: I'm not sure this is what Rob meant by checking the review exists, I guess it works but it's really difficult to read

File: __tests__/app.test.js [hunk 12]
```diff
@@ -139,18 +165,69 @@ describe("GET /api/reviews/:review_id", () => {
...
+    return request(app)
+      .get("/api/reviews/2/comments")
+      .then(({ body }) => {
+        body.comments.forEach((comment) => {
+          expect(comment).toMatchObject({
+            comment_id: expect.any(Number),
+            body: expect.any(String),
+            votes: expect.any(Number),
+            author: expect.any(String),
+            review_id: expect.any(Number),
```
Comment (roisinoreilly) [inline __tests__/app.test.js:205]: again this needs to check for the specific review number

**Review by roisinoreilly — APPROVED** [review]
It's really important to remove all .onlys in your code, you could have written functionality that breaks all the rest of your tests and you wouldn't know. There's still a few changes that need implementing but after you've done that you can merge
